src/number_theory/number_field.lean

```diff
@@ -0,0 +1,118 @@
+/-
+Copyright (c) 2021 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ashvni Narayanan, Anne Baanen
+-/
+
+import algebra.field
+import data.rat.basic
+import ring_theory.algebraic
+import ring_theory.dedekind_domain
+import ring_theory.integral_closure
+import ring_theory.polynomial.rational_root
+
+/-!
+# Number fields
+This file defines a number field and the ring of integers corresponding to it.
+
+## Main definitions
+ - `number_field` defines a number field as a field which has characteristic zero and is finite
+    dimensional over ℚ.
+ - `ring_of_integers` defines the ring of integers (or number ring) corresponding to a number field
+    as the integral closure of ℤ in the number field.
+
+## Implementation notes
+The definitions that involve a field of fractions choose a canonical field of fractions,
+but are independent of that choice.
+
+## References
+* [D. Marcus, *Number Fields*][marcus1977number]
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+* [P. Samuel, *Algebraic Theory of Numbers*][samuel1970algebraic]
+
+## Tags
+number field, ring of integers
+-/
+
+/-- A number field is a field which has characteristic zero and is finite
+dimensional over ℚ. -/
+class number_field (K : Type*) [field K] : Prop :=
+[to_char_zero : char_zero K]
+[to_finite_dimensional : finite_dimensional ℚ K]
+
+open function
+open_locale classical big_operators
+
+namespace number_field
+variables (K : Type*) [field K] [number_field K]
+
+-- See note [lower instance priority]
+attribute [priority 100, instance] number_field.to_char_zero number_field.to_finite_dimensional
+
+protected lemma is_algebraic : algebra.is_algebraic ℚ K := algebra.is_algebraic_of_finite
+
+/-- The ring of integers (or number ring) corresponding to a number field
+is the integral closure of ℤ in the number field. -/
+@[nolint unused_arguments] -- There are multiple definitions of rings of integers.
+def ring_of_integers := integral_closure ℤ K
+
+namespace ring_of_integers
+
+variables {K}
+
+lemma is_integral_coe (x : ring_of_integers K) : is_integral ℤ (x : K) :=
+x.2
+
+variables (K)
+
+lemma char_zero.of_algebra {R : Type*} (S : Type*) [comm_semiring R] [semiring S] [algebra R S]
+  [char_zero S] : char_zero R :=
+⟨begin
+  suffices : injective (algebra_map R S ∘ coe),
+  { exact injective.of_comp this },
+  convert char_zero.cast_injective,
+  ext n,
+  rw [function.comp_app, ← (algebra_map ℕ _).eq_nat_cast, ← ring_hom.comp_apply,
+      ring_hom.eq_nat_cast],
+  all_goals { apply_instance }
+end⟩
```

Comment:
This presumably belongs somewhere else, right?

Reply:
Good catch. I must have forgotten to actually proofread this file yesterday, these are quite embarrassing errors.